spec.html

```diff
@@ -39097,6 +38930,256 @@
+
+    <emu-annex id="sec-scriptevaluationjob" aoid="ScriptEvaluationJob">
+      <h1>ScriptEvaluationJob ( _sourceText_ )</h1>
+      <p>A ScriptEvaluationJob with parameter _sourceText_ parses, validates, and evaluates _sourceText_ as a |Script|.</p>
+      <emu-alg>
+        1. Assert: _sourceText_ is an ECMAScript source text (see clause <emu-xref href="#sec-ecmascript-language-source-code"></emu-xref>).
+        1. Let _realm_ be the current Realm Record.
+        1. Let _s_ be ParseScript(_sourceText_, _realm_, *undefined*).
+        1. If _s_ is a List of errors, then
+          1. Perform ReportErrors(_s_).
+          1. Return NormalCompletion(*undefined*).
+        1. Return ? ScriptEvaluation(_s_).
+      </emu-alg>
+    </emu-annex>
+
+    <emu-annex id="sec-toplevelmoduleevaluationjob" aoid="TopLevelModuleEvaluationJob">
+      <h1>TopLevelModuleEvaluationJob ( _sourceText_, _filename_ )</h1>
+      <p>A TopLevelModuleEvaluationJob with parameters _sourceText_ and _filename_ is a job that parses, validates, and evaluates _sourceText_ as a |Module|.</p>
+      <emu-alg>
+        1. Assert: _sourceText_ is an ECMAScript source text (see clause <emu-xref href="#sec-ecmascript-language-source-code"></emu-xref>).
+        1. Let _realm_ be the current Realm Record.
+        1. Let _m_ be ParseModule(_sourceText_, _realm_, _filename_).
+        1. If _m_ is a List of errors, then
+          1. Perform ReportErrors(_m_).
+          1. Return NormalCompletion(*undefined*).
+        1. Perform ? _m_.ModuleDeclarationInstantiation().
+        1. Assert: All dependencies of _m_ have been transitively resolved and _m_ is ready for evaluation.
+        1. Return ? _m_.ModuleEvaluation().
+      </emu-alg>
+      <emu-note>
+        <p>The implementation might parse a _sourceText_ as a |Module|, analyze it for Early Error conditions, and instantiate it prior to the execution of the TopLevelModuleEvaluationJob for that _sourceText_. It might also resolve, pre-parse and pre-analyze, and pre-instantiate module dependencies of _sourceText_. However, the reporting of any errors detected by these actions will be deferred until the TopLevelModuleEvaluationJob is actually executed.</p>
+      </emu-note>
+    </emu-annex>
+
+    <emu-annex id="sec-example-module-resolution">
+      <h1>Module Resolution</h1>
+      <p>This implementation uses the [[HostDefined]] field of Source Text Module Records to store the filename from which a module was retrieved. It also maintains a List of ModuleMap Records, known as the module map. This helps enforce the idempotency requirements of HostResolveImportedModule. ModuleMap Records have the fields described here:</p>
```

Comment:
An alternative solution here to avoid having to use `Filename` (which sounds like some alien to this specification) is to create a caching mechanism based on the module specifier. We played around with this recently, and seems to be sufficient to maintain the idempotency requirements. It will be very similar to what you have here, but just mapping between the module specifier (per dep) and the corresponding module record, and using the Filename only if the record is not found in the map.

This might be important for the future when we introduce a way to create other kind of records, in which case we might want to pre-link them before they are instantiated.